### PR TITLE
refactor(api): refactor smoothie and module response parsing

### PR DESCRIPTION
**Pull request by amitlissack**
# Overview

Reduces duplication of much of the smoothie response parsing across smoothie and module drivers. 

Much of our parsing was inefficient: splitting strings twice at `:`, once for each side.

# Changelog

- added methods to `opentrons.drivers.utils`:
`parse_key_values` to convert from `"X:1 Y:2"` to `{"X":1, "Y":2}`
`parse_temperature_response` to parse a `Temperature` object (thermocycler and tempdeck)
`parse_plate_temperature_response`  to parse a `PlateTemperature` object (thermocycler)
`parse_optional_number` to parse an int that maybe "none"
`parse_number` - to parse a number
- removed methods from `opentrons.drivers.utils`:
`parse_number_from_substring` - deprecated  by `parse_key_values`
`parse_key_from_substring` - deprecated  by `parse_key_values`
- removed methods from `opentrons.drivers.magdeck.driver`:
`_parse_string_value_from_substring` - deprecated  by `parse_key_values`
`_parse_number_from_substring` - deprecated  by `parse_key_values`
`_parse_key_from_substring` - deprecated  by `parse_key_values`
`_parse_device_information` - duplicated by `driver.utils.parse_device_information`
- modify modules and smoothie drivers to use the new functions.

# Review requests

This needs to be tested to ensure that smoothie, magdeck, tempdeck, and thermocycler behaviour is unchanged. 

# Risk assessment

**Very high**. Affects parsing smoothie and module responses.

### CODE REVIEW

File: api/src/opentrons/drivers/temp_deck/driver.py [hunk 1]
```diff
@@ -189,26 +190,15 @@ def start_set_temperature(self, celsius) -> str:
...
-        self._temperature.update({'target': celsius})
-        return ''
-
-    def update_temperature(self, default=None) -> str:
+    def update_temperature(self) -> str:
         if self._update_thread and self._update_thread.is_alive():
-            updated_temperature = default or self._temperature.copy()
-            self._temperature.update(updated_temperature)
+            pass
+            # TODO AL 20210320 - I don't know what this for. default was never used.
```
Comment (amitlissack) [inline api/src/opentrons/drivers/temp_deck/driver.py]: Anyone know about this?

Comment (sanni-t) [reply api/src/opentrons/drivers/temp_deck/driver.py]: Not sure what it was originally meant for but from [this PR](https://github.com/Opentrons/opentrons/pull/1530/files) it looks like `default` was supposed to be reassigned with `self._temperature`'s value before line 200. Looks like it could have meant to differentiate between simulating & non-simulating behavior.

tl;dr: I think it's okay to remove `default`

File: api/src/opentrons/drivers/mag_deck/driver.py [hunk 2]
```diff
@@ -58,100 +60,21 @@ class MagDeckError(Exception):
...
-    if 'Z' not in distance_string and 'height' not in distance_string:
-        raise ParseError(err_msg)
-    return _parse_number_from_substring(  # type: ignore
-        distance_string.strip())          # (preconditions checked above)
+    data = utils.parse_key_values(distance_string)
+    val = data.get('Z', data.get('height'))
+    if val is None:
+        raise utils.ParseError(
+            f'Unexpected argument to _parse_distance_response: '
+            f'{distance_string}')
```
Comment (SyntaxColoring) [inline api/src/opentrons/drivers/mag_deck/driver.py]: Should we instead rely on static type checking for this?

File: api/src/opentrons/drivers/utils.py [hunk 3]
```diff
@@ -10,6 +13,9 @@
 TC_GCODE_ROUNDING_PRECISION = 2
 
 
+KEY_VALUE_REGEX = re.compile(r"((?P<key>\S+):(?P<value>\S+))")
+
+
 class ParseError(Exception):
```
Comment (SyntaxColoring) [inline api/src/opentrons/drivers/utils.py:19]: Doesn't necessarily have to happen in this PR, but all uses of `ParseError` appear to include the problematic source string in the error message. Does it make sense to make this type more structured, like:

```python
class ParseError(Exception):
    def __init__(error_message: str, parse_source: str):
        self.error_message = error_message
        self.parse_source = parse_source
        super().__init__(f"ParseError(error_message={repr(error_message)}, parse_source={repr(parse_source)})")
```

To deduplicate the error message string formatting.

File: api/src/opentrons/drivers/utils.py [hunk 4]
```diff
@@ -29,96 +35,86 @@ def parse_string_value_from_substring(substring) -> str:
...
+def parse_key_values(value: str) -> Dict[str, str]:
+    """Convert string in the format:
+        'key1:value1 key2:value2'
+        to dict
+        {'key1': 'value1', 'key2': 'value2'}
+    """
+    if not value or not isinstance(value, str):
+        raise ParseError(
+            f'Unexpected argument to parse_key_values: {value}'
+        )
```
Comment (SyntaxColoring) [inline api/src/opentrons/drivers/utils.py]: Ditto earlier typechecking comment.

File: api/src/opentrons/drivers/utils.py [hunk 5]
```diff
@@ -29,96 +35,86 @@ def parse_string_value_from_substring(substring) -> str:
...
+def parse_optional_number(value: str, rounding_val) -> Optional[float]:
+    """Convert number to float. 'none' will be converted to None"""
+    return None if value == "none" else parse_number(value, rounding_val)
+
+
+def parse_number(value: str, rounding_val: int) -> float:
+    """Convert string to float."""
+    try:
+        return round(float(value), rounding_val)
+    except (ValueError, IndexError, TypeError, AttributeError):
```
Comment (SyntaxColoring) [inline api/src/opentrons/drivers/utils.py]: I *think* the only thing that can fail here is `value` not being formatted like a float, in which case `float()` raises just `ValueError`. I don't see how `IndexError`, `TypeError`, or `AttributeError` could be raised.

File: api/src/opentrons/drivers/utils.py [hunk 6]
```diff
@@ -29,96 +35,86 @@ def parse_string_value_from_substring(substring) -> str:
...
+        {'key1': 'value1', 'key2': 'value2'}
+    """
+    if not value or not isinstance(value, str):
+        raise ParseError(
+            f'Unexpected argument to parse_key_values: {value}'
+        )
+
+    res = {g.groupdict()['key']: g.groupdict()['value']
+           for g in KEY_VALUE_REGEX.finditer(value)}
+    return res
```
Comment (SyntaxColoring) [inline api/src/opentrons/drivers/utils.py]: I don't know if we care, but in a string like

```
'a:b c d:e'
```

This will return `{'a':'b', 'd':'e'}`, silently ignoring `c`.

File: api/tests/opentrons/drivers/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        ['serial:serial_v model:m version:123-2 ',
+         {'version': '123-2', 'serial': 'serial_v', 'model': 'm'}],
+        ['   serial:serial_v model:m    version:123-2   ',
+         {'version': '123-2', 'serial': 'serial_v', 'model': 'm'}]
+    ]
+)
+def test_parse_device_information_success(
+        input_str: str, expected_result: Dict[str, str]) -> None:
+    """Test parse device information."""
+    assert utils.parse_device_information(input_str) == expected_result
```
Comment (SyntaxColoring) [inline api/tests/opentrons/drivers/test_utils.py]: Should we also test the current (I think) behavior that extra `k:v` pairs are silently ignored?

```python
['serial:serial_v extra_k:extra_v model:m version:123-2',
 {'version': '123-2', 'serial': 'serial_v', 'model': 'm'}]
```

I don't know whether or not this is intended/important.

**Review by SyntaxColoring — APPROVED** [review]
Cool!

Various comments but, but overall, this PR looks good to me if you don't want to address these comments now.

File: api/src/opentrons/drivers/mag_deck/driver.py (hunk 2 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/mag_deck/driver.py]: How? Get returns an `Optional[Any]`.

File: api/src/opentrons/drivers/utils.py (hunk 6 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/utils.py]: I don't care about it.

File: api/src/opentrons/drivers/utils.py (hunk 5 above)
Comment (amitlissack) [reply api/src/opentrons/drivers/utils.py]: Those are remnants from the past probably.

File: api/tests/opentrons/drivers/test_utils.py (hunk 7 above)
Comment (amitlissack) [reply api/tests/opentrons/drivers/test_utils.py]: Sure. Why not.

File: api/src/opentrons/drivers/mag_deck/driver.py (hunk 2 above)
Comment (SyntaxColoring) [reply api/src/opentrons/drivers/mag_deck/driver.py]: My mistake. I misread this. I thought `val` came in as a function argument.

File: api/tests/opentrons/drivers/test_utils.py (hunk 7 above)
Comment (amitlissack) [reply api/tests/opentrons/drivers/test_utils.py]: Thanks. I realize that the caller was responsible for ignoring extra keys. I changed it to omit extra keys.

File: api/src/opentrons/drivers/utils.py [hunk 8]
```diff
@@ -17,10 +17,15 @@
 
 
 class ParseError(Exception):
-    pass
+    def __init__(self, error_message: str, parse_source: str) -> None:
+        self.error_message = error_message
+        self.parse_source = parse_source
+        super().__init__(
+            f"ParseError(error_message={error_message}, parse_source={parse_source})"
```
Comment (SyntaxColoring) [inline api/src/opentrons/drivers/utils.py:24]:
```suggestion
            f"ParseError(error_message={repr(error_message)}, parse_source={repr(parse_source)})"
```

I like `repr()` for things like this because, when it gets printed as part of an exception stack trace, it will automatically handle unprintable characters and wrap the string in quotation marks.

```python
m = '\x00foo'
raise Exception(f'm={m}') # Exception: m=foo
raise Exception(f'm={repr(m)}') # Exception: m='\x00foo'
```

Definitely not standard practice or anything—I don't know if anyone else does this—so, up to you.

File: api/tests/opentrons/drivers/test_utils.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+@pytest.mark.parametrize(
+    argnames=['input_str', 'expected_result'],
+    argvalues=[
+        ['T:none C:none H:none',
+         PlateTemperature(target=None, current=None, hold=None)],
+        ['T:321.4 C:45 H:123.222',
+         PlateTemperature(target=321.4, current=45, hold=123.22)],
+        ['T:-44.2442 C:-22.233 H:0',
+         PlateTemperature(target=-44.24, current=-22.23, hold=0)],
+    ]
```
Comment (sanni-t) [inline api/tests/opentrons/drivers/test_utils.py:87]: Since there are times (while debugging & maybe in a future thermocycler firmware version) where we want to make sure that extra fields in temp status are ignored, it might be good to test for that too..

```suggestion
         ['T:-44.2442 C:-22.233 H:0 foo:bar',
          PlateTemperature(target=-44.24, current=-22.23, hold=0)],
    ]
```

File: api/tests/opentrons/drivers/test_utils.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    argnames=['input_str'],
+    argvalues=[
+        ['T:not_a_float C:123 H:321'],
+        ['C:not_a_float T:123 H:321'],
+        ['H:not_a_float C:123 T:321'],
+        [''],
+    ]
+)
+def test_parse_plate_temperature_response_failure(input_str: str) -> None:
+    """It should faile to parse plate temperature response"""
```
Comment (sanni-t) [inline api/tests/opentrons/drivers/test_utils.py]:
```suggestion
    """It should fail to parse plate temperature response"""
```

**Review by sanni-t — APPROVED** [review]
Looks great!
Just a thought about the thermocycler test

**Review by nusrat813 — APPROVED** [review]
Tested - LGTM
